source/agora/consensus/Quorum.d

```diff
@@ -0,0 +1,192 @@
+/*******************************************************************************
+
+    Contains the quorum generator algorithm.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.Quorum;
+
+import agora.common.Config;
+import agora.common.crypto.Key;
+import agora.common.Set;
+import agora.common.Types;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.UTXOSet;
+
+import std.algorithm;
+import std.array;
+import std.exception;
+import std.format;
+import std.math;
+import std.range;
+import std.string;
+
+/*******************************************************************************
+
+    Build the quorum configuration for the given public key and the staked
+    enrollments.
+
+    Params:
+        node_key = the key of the node for which to generate the quorum
+        enrolls = the array of registered enrollments
+        finder = UTXO finder delegate
+
+    Returns:
+        the generated quorum config
+
+*******************************************************************************/
+
+public QuorumConfig buildQuorumConfig ( const ref PublicKey node_key,
+    in Enrollment[] enrolls, in UTXOFinder finder )
+{
+    Set!PublicKey all_nodes = getPublicKeys(enrolls, finder);
+
+    QuorumConfig quorum;
+    all_nodes.each!(node => quorum.nodes ~= node);
+    quorum.nodes.sort();
+    quorum.threshold = quorum.nodes.length;
+
+    return quorum;
+}
+
+///
+unittest
+{
+    // test up to 1024 nodes
+    foreach (num_nodes; iota(1, 10).map!(n => 2 ^^ n))
```

Comment:
![That's how I felt](https://i.imgflip.com/41vsye.jpg)

Reply:
Yeah sometimes I find myself abusing ranges just to write something clever. I could have just written 2, 4, 8, 16, 32, 64, 128.. etc I guess.